spec/ics-2-consensus-requirements/README.md

```diff
@@ -0,0 +1,89 @@
+
+Definitions:
+
+1. Direct children have their parent as `p` and are verifiable with their parent.
+2. The height of a block is the step of referring `p` required to get to `nil`
+
+Requirements:
+
+1. Blocks have only one direct child
+If a blockchain has deterministic safety(as opposed to probabilistic safety in Nakamoto consensus), then there cannot be more than one child for each block. In tendermint, this assumption breaks when +1/3 validators are malicious, producing multiple blocks those are direct children of a single block and all of them are verifiable with the parent. It makes conflicting packets delivered from the failed chain, so for the chains who are receiving from this chain, **fraud proof #10** mechanism should be applied in this case. However, he failed chain itself also need to be recovered and reconnected again. It will be covered in **byzantine recovery strategies #6**.
+
+2. Blocks have at least one direct child
+There is at least one direct child for all blocks, meaning that the lightclient logic can proceed the blocks one by one even in the worst case. If it not satisfied then there can be a point where the lightclient stops and cannot proceed, which halts IBC connection unexpectedly. This also can be violated when the blocks are restarted out of the consensus, for example in Tendermint, (1/2 < validators < 2/3) forked out the chain. This also will be covered in **byzantine recovery strategies**.
+
+3. If a block verifies another block then it is a descendant of the block
+Lightclient should not verify packets which are not in its chain. If the verifier returns true for a block that is not a descendant, it simply means that there is an error in the lightclient logic.
+
+4. Block can have a state only if it is a valid transition from the parent's(see the next paragraph)
+Block cannot have a state which is not a transition from its parent. Since the state is only defined for IBC logic, this only means that each block are running IBC-compatible application logic(does not care about the other modules' failure, including the handlers for the packets).
+
+These requirements allow channels to work safely without concerning about double spending attack. 
+If a block is submitted to the chain(and optionally passed some challenge period for fraud proof), then it is assured that the packet is finalized so the application logic can process it.
+
+### Technical Specification
+
+In a real world, `Block`s do not have to contain the full state data; it can contain an aggregated data set identifier, for example a merkle root, where the actual data latter given can be proven that it is a member of that set. If we don't consider about the data availability problem, this partial blocks, which is be called `Header`s, still satisfies `Block`.
+
+Following functions exist over `Header`s.
+
+* `height : Header -> int`
```

Comment:
What do we need this for? Bisection?

Reply:
Yes, but maybe we can make `verify` to handle it(in `Maybe<LightClientProof>`)

Reply:
I think that would be cleaner.

Reply:
Height is still needed: the verifier needs to get the trusted header in prior to verification which requires height information.

Reply:
I don't think I understand, let's discuss this.